Add new “Scripts and Event Handling” guide (#2102)
withastro/docs: src/pages/en/guides/client-side-scripts.md

## Using `<script>` in Astro

In `.astro` files, you can add client-side JavaScript by adding one (or more) `<script>` tags.

In this example, adding the `<Hello />` component to a page will log a message to the browser console.

```astro title="src/components/Hello.astro"
<h1>Welcome, world!</h1>

<script>
  console.log('Welcome, browser console!');
</script>
```

### Script bundling

By default, `<script>` tags are processed by Astro.

- Any imports will be bundled, allowing you to import local files or Node modules.
- The processed script will be injected into your page’s `<head>` with [`type="module"`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
- TypeScript is fully supported, including importing TypeScript files
- If your component is used several times on a page, the script will only be included once.

```astro title="src/components/Example.astro"
<script>
  // Processed! Bundled! TypeScript-supported!

>   // ESM imports work, even to npm packages.

Reviewer: Flagging that this, in conjunction with/as a contrast to the inline script, needs some work.

Author: ```suggestion
  // Importing local scripts and Node modules works.
```

</script>
```

To avoid bundling the script, you can use the `is:inline` attribute.

```astro title="src/components/InlineScript.astro" "is:inline"
<script is:inline>
  // Will be rendered into the HTML exactly as written!

>   // ESM imports will not be resolved relative to the file.

Reviewer: Reminding us that we are not sold on this!

Author: ```suggestion
  // Local imports are not resolved and will not work.
```

Author: Another option could even be just to remove this line wholesale.

We’ve said “rendered as written” and that the script isn’t bundled, so might be sufficient.

```suggestion
```

</script>
```

:::note

> Adding `type="module"` or any other attribute to a `<script>` tag will disable Astro's default bundling behavior, treating the tag as if it had an `is:inline` directive.

Author: You _can_ add `src` and take advantage of bundling (as we show below) so tweaking this:

```suggestion
Adding `type="module"` or any attribute other than `src` to a `<script>` tag will disable Astro's default bundling behavior, treating the tag as if it had an `is:inline` directive.
```

:::

📚 See our [directives reference](/en/reference/directives-reference/#script--style-directives) page for more information about the directives available on `<script>` tags.

### Script loading

You may want to write your scripts as separate `.js`/`.ts` files or need to reference an external script on another server. You can do this by referencing these in a `<script>` tag’s `src` attribute.

#### Import local scripts

**When to use this:** If your script lives inside of `src/`.

> Astro will build, optimize, and add these scripts to the page for you, following its [script bundling rules](#script-bundling-in-astro).

Author: ```suggestion
Astro will build, optimize, and add these scripts to the page for you, following its [script bundling rules](#script-bundling).
```

```astro title="src/components/LocalScripts.astro"
<!-- relative path to script in `src/` -->
<script src="./local-script.js"></script>

<!-- also works for local TypeScript files -->
<script src="./script-with-types.ts"></script>
```

#### Load external scripts

**When to use this:** If your JavaScript file lives inside of `public/` or on a CDN.

This approach skips the JavaScript processing, bundling, and optimizations that are provided by Astro when you import scripts as described above.

```astro title="src/components/ExternalScripts.astro" "is:inline"
<!-- absolute path to a script at `public/my-script.js` in your project -->
<script is:inline src="/my-script.js"></script>

<!-- full URL to a script on a remote server -->
<script is:inline src="https://analytics.example.com/script.js"></script>
```

## Common script patterns

### Handle `onclick` and other events

> Some UI frameworks use custom syntax for event handling like `onclick={...}` (React/Preact) or `@click="..."` (Vue). Astro components follow web standards and recommend using [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) in a `<script>` tag to handle user interactions.

Author: Here’s what I went with for now:

> Some UI frameworks use custom syntax for event handling like `onClick={...}` (React/Preact) or `@click="..."` (Vue). This is not possible in Astro components, which follow standard HTML syntax more closely.
> 
> Instead, you can use [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) in a `<script>` tag to handle user interactions.

Do you think that’s OK?

```astro title="src/components/AlertButton.astro"
<button data-alert>Click me!</button>

<script>
  // Find all buttons with the `data-alert` attribute on the page.
  const buttons = document.querySelectorAll('button[data-alert]');

  // Handle clicks on each button.
  buttons.forEach((button) => {
    button.addEventListener('click', () => {
      alert('Button was clicked!');
    });
  });
</script>

> ```

Reviewer: If this example works with a css `class=""` instead of a `data-` attribute, I'm *tempted* to show that here instead. (Not opposed to what's here, just thinking...)

My thinking: 
- CSS classes might be more familiar, and might already be in use for an element anyway.
- You have a nice introduction/explanation to `data-`  attributes in the "using frontmatter variables" section. So, it seems weird to be using them here, then introducing them at the bottom of the page.

We *could* move that introduction up here,  but I think this is a simple enough (and often searched!) example that we don't need to clutter what's not very complicated here.

In which case I might even add to the line above with something like:

> You can add [`addEventListeners`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) to elements identified by class, id, attribute  etc. in a `<script>` tag to handle user interactions.

Author: Agree — `data-*` attributes are often recommended for this because they help you separate styling and behaviour, but `class` examples are more common, and showing the data attribute later means we can cover both bases.

````suggestion
```astro title="src/components/AlertButton.astro"
<button class="alert">Click me!</button>

<script>
  // Find all buttons with the class `alert` attribute on the page.
  const buttons = document.querySelectorAll('button.alert');

  // Handle clicks on each button.
  buttons.forEach((button) => {
    button.addEventListener('click', () => {
      alert('Button was clicked!');
    });
  });
</script>
```
````

:::note

> Because this script will run once even if we add many instances of `<AlertButton />` to a page, we use `querySelectorAll` to get all instances of our component and add listeners to each one instead of using `querySelector`, which would only return the first button it found.

Reviewer 2: small suggestion, feel free to ignore
```suggestion
Because this script will be shipped only once even if we add many instances of `<AlertButton />` to a page, we use `querySelectorAll` to get all instances of our component and add listeners to each one instead of using `querySelector`, which would only return the first button it found.
```

Author: Oops, this got overridden by Sarah’s rewrite here. The new wording is “Scripts are bundled and run only once.” Do you think “bundled” is enough to convey the same concept as “shipped” here?

> Because this script will run once even if we add many instances of `<AlertButton />` to a page, we use `querySelectorAll` to get all instances of our component and add listeners to each one instead of using `querySelector`, which would only return the first button it found.

Author: Not sure how different these are in messaging, but I do like your wording better. Tweaked ever so slightly:

```suggestion
If you have multiple `<AlertButton />` components on a page, Astro will not run the component's script multiple times. Scripts are bundled and run only once. Using `querySelectorAll` ensures that this script attaches the event listener to every button with the `alert` class found on the page.
```

:::

### Web components with custom elements

> [Custom elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) provide a standard API that allows you to define your own HTML elements with custom behavior. Defining a custom element in a `.astro` component can be a good way to build interactivity without reaching for a UI framework library.

Author: 👍 just removing a stray space:

```suggestion
You can create your own HTML elements with custom behavior using the Web Components standard. Defining a [custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) in a `.astro` component allows you to build interactive components without needing a UI framework library.
```

> [Custom elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) provide a standard API that allows you to define your own HTML elements with custom behavior. Defining a custom element in a `.astro` component can be a good way to build interactivity without reaching for a UI framework library.

Reviewer: Haha... my space bar was stiiiiiicking last night! It was a struggle to fight against either no spaces, or spaces appearing before my eyes. Thanks for the catch!

In this example, we define a new `<astro-heart>` HTML element that tracks how many times you click the heart button and updates the `<span>` with the latest count.

```astro title="src/components/AstroHeart.astro"
<!-- Wrap the component elements in our custom element “astro-heart”. -->
<astro-heart>
  <button aria-label="Heart">💜</button> × <span>0</span>
</astro-heart>

<script>
  // Define the behaviour for our new type of HTML element.
  class AstroHeart extends HTMLElement {
    constructor() {
			super();
      let count = 0;

      const heartButton = this.querySelector('button');
      const countSpan = this.querySelector('span');

      // Each time the button is clicked, update the count.
			heartButton.addEventListener('click', () => {
        count++;
        countSpan.textContent = count;
      });
		}
  }

  // Tell the browser to use our AstroHeart class for <astro-heart> elements.
  customElements.define('astro-heart', AstroHeart);
</script>
```

> This is similar to how you might handle user input without a custom element. But, instead of using `document.querySelector()`, we used `this.querySelector()` to get DOM elements. `this.querySelector()` only searches within the current custom element instance, so it’s easier to work with the children of one component instance at a time. Because the browser understands that this code is for a custom element, it will automatically run it for you each time it finds `<astro-heart>` on the page.

Reviewer: Flagging as needs work!  😅 

I'm not sure *I* get why this is different than the previous case, other than "your browser understands that this code is for a custom element"... so in that spirit, I'm going to try something (that you will correct later) like...
```suggestion
 Unlike scripts that are bundled and only run once, the script for `<astro-heart>` will be run by the browser each time it finds this custom element on the page. The browser understands that `this.querySelector()` only searches within the current custom element instance to get DOM elements. So, even if you have multiple `<AstroHeart />` components, its script does not have to query for/add event listenters to all instances  of its child elements.
```

Author: Yeah — I wasn’t happy with this either. Here’s a second attempt.

```suggestion
There are two advantages to using a custom element here:

1. You can use `this.querySelector()` to get DOM elements instead of searching the whole page using `document.querySelector()`. `this.querySelector()` only searches within the current custom element instance, making it easier to work with the children of one component instance at a time.

2. The browser will run our custom element’s `constructor()` method each time it finds `<astro-heart>` on the page. This means your code only needs to handle one component at a time instead of setting up all components on the page at once.
```

Will commit this so I can see it better in context and await feedback.

> 

Reviewer 2: if we were to use this component multiple times, will the javascript be shipped only once, or once per instance?

📚 You can learn more about custom elements in [web.dev’s Reusable Web Components guide](https://web.dev/custom-elements-v1/) and [MDN’s introduction to custom elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements).

> ### Pass frontmatter variables to scripts

Reviewer 2: can we include the `define:vars` example here too 🙏
```suggestion
### Pass frontmatter variables to scripts with web components
```

In Astro components, the code in [the frontmatter](/en/core-concepts/astro-components/#the-component-script) between the `---` fences runs on the server and is not available in the browser. To send variables from the server to the client, we need a way to store our variables and then read them when JavaScript runs in the browser.

> One way to do this is to use [`data-*` attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) to store the value of variables in your HTML output. Scripts, including custom elements, can then read these once your HTML loads in the browser.
> 
> In this example component, a `message` prop is stored in a `data-message` attribute. Data attributes are available in JavaScript using an element’s `dataset` property, so the custom element logic can read `this.dataset.message` and get the value of the prop in the browser.

Reviewer: ```suggestion
One way to do this is to use [`data-*` attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) to store the value of variables in your HTML output. Scripts, including custom elements, can then read `this.dataset.*` once your HTML loads in the browser.

In this example component, a `message` prop is stored in a `data-message` attribute and is available in the browser at `this.dataset.message`.
```
Is this true? It's more concise.  😄

Author: It’s not really true, no 😆 

An element’s data attributes are available in the `dataset` object _for that element_.

So, for example:

```html
<p data-marquee>Thanksgiving Offer: chickens going cheep!</p>

<script>
const p = document.querySelector('p');
// p.dataset has a marquee property
p.dataset.marquee;
</script>
```

In our case, we can use `this.dataset` because `this` is the root node in a custom element, but `this.dataset` is not simply “available in the browser.”

Would this be an improvement do you think?

```suggestion
One way to do this is to use [`data-*` attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) to store the value of variables in your HTML output. Scripts, including custom elements, can then read these using an element’s `dataset` property once your HTML loads in the browser.

In this example component, a `message` prop is stored in a `data-message` attribute, so the custom element can read `this.dataset.message` and get the value of the prop in the browser.
```

Author: Selfishly going to commit my variant so I can reread the full thing.

#### Import local scripts

**When to use this:** If your script lives inside of `src/`.

Astro will build, optimize, and add these scripts to the page for you, following its [script bundling rules](#script-bundling).

```astro title="src/components/LocalScripts.astro"
<!-- relative path to script at `src/scripts/local.js` -->
<script src="../scripts/local.js"></script>

<!-- also works for local TypeScript files -->
<script src="./script-with-types.ts"></script>
```

#### Load external scripts

**When to use this:** If your JavaScript file lives inside of `public/` or on a CDN.

To load scripts outside of your project's `src/` folder, include the `is:inline` directive. This approach skips the JavaScript processing, bundling, and optimizations that are provided by Astro when you import scripts as described above.

```astro title="src/components/ExternalScripts.astro" "is:inline"
<!-- absolute path to a script at `public/my-script.js` -->
<script is:inline src="/my-script.js"></script>

<!-- full URL to a script on a remote server -->
<script is:inline src="https://my-analytics.com/script.js"></script>
```

## Common script patterns

### Handle `onclick` and other events

> Some UI frameworks use custom syntax for event handling like `onClick={...}` (React/Preact) or `@click="..."` (Vue). This is not possible in Astro components, which follow standard HTML syntax more closely.
> 
> Instead, you can use [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) in a `<script>` tag to handle user interactions.

Author: ```suggestion
Some UI frameworks use custom syntax for event handling like `onClick={...}` (React/Preact) or `@click="..."` (Vue). Astro follows standard HTML more closely and does not use custom syntax for events.

Instead, you can use [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) in a `<script>` tag to handle user interactions.
```

Reviewer: I can get behind this! LGTM!

```astro title="src/components/AlertButton.astro"
<button class="alert">Click me!</button>

<script>
  // Find all buttons with the `alert` class on the page.
  const buttons = document.querySelectorAll('button.alert');

  // Handle clicks on each button.
  buttons.forEach((button) => {
    button.addEventListener('click', () => {
      alert('Button was clicked!');
    });
  });
</script>
```

:::note

> If you have multiple `<AlertButton />` components on a page, Astro will not run the component's script multiple times. Scripts are bundled and run only once. Using `querySelectorAll` ensures that this script attaches the event listener to every button with the `alert` class found on the page.

Author: Slight rework to try to address Kevin’s comment.

```suggestion
If you have multiple `<AlertButton />` components on a page, Astro will not run the script multiple times. Scripts are bundled and only included once per page. Using `querySelectorAll` ensures that this script attaches the event listener to every button with the `alert` class found on the page.
```

Reviewer: Approve!

In this example, we define a new `<astro-heart>` HTML element that tracks how many times you click the heart button and updates the `<span>` with the latest count.

```astro title="src/components/AstroHeart.astro"
<!-- Wrap the component elements in our custom element “astro-heart”. -->
<astro-heart>
  <button aria-label="Heart">💜</button> × <span>0</span>
</astro-heart>

<script>
  // Define the behaviour for our new type of HTML element.
  class AstroHeart extends HTMLElement {
    constructor() {
			super();
      let count = 0;

      const heartButton = this.querySelector('button');
      const countSpan = this.querySelector('span');

      // Each time the button is clicked, update the count.
			heartButton.addEventListener('click', () => {
        count++;
        countSpan.textContent = count;
      });
		}
  }

  // Tell the browser to use our AstroHeart class for <astro-heart> elements.
  customElements.define('astro-heart', AstroHeart);
</script>
```

There are two advantages to using a custom element here:

> 1. You can use `this.querySelector()` to get DOM elements instead of searching the whole page using `document.querySelector()`. `this.querySelector()` only searches within the current custom element instance, making it easier to work with the children of one component instance at a time.

Reviewer: ```suggestion
1. You can use `this.querySelector()` to get DOM elements instead of searching the whole page using `document.querySelector()`. `this.querySelector()` only searches within the current custom element instance, making it easier to target the children of one component instance at a time.
```
?? I feel like this might relate back to "querying" a bit better? Just a thought.

Author: Is “target“ verging on jargon? I could go either way but leaning “work with” as descriptive of the fact you want to _do_ things with them.

Reviewer: I was updating this comment as you were replying! I can go back to "work with"

> 2. The browser will run our custom element’s `constructor()` method each time it finds `<astro-heart>` on the page. This means your code only needs to handle one component at a time instead of setting up all components on the page at once.

Reviewer: ```suggestion
2. Although a `<script>` is only run once, the browser will run our custom element’s `constructor()` method each time it finds `<astro-heart>` on the page. This means you can safely write code for one component at a time, even if you intend to use this components multiple times on a page.
```
?? I do like setting up a callback/contrast to before when we said you had to write for multiple instances, so whatever the correct opening clause is, I think one would work here.

Another possible take...
> Unlike a script that is bundled and only run once, any custom element's `constructor()` method will be run ...

Also, since this is framed as an "advantage", we could consider something a little more direct re: the code author "YOU CAN DO THIS..."

For your consideration!

Author: Excellent! Tiny tweaks before committing:

```suggestion
2. Although a `<script>` only runs once, the browser will run our custom element’s `constructor()` method each time it finds `<astro-heart>` on the page. This means you can safely write code for one component at a time, even if you intend to use this component multiple times on a page.
```
